src/parser/jobs/drg/modules/DrgBuffWindow.tsx

```diff
@@ -0,0 +1,210 @@
+		if (action && action.onGcd) {
+			this.lastGcd = action
+		}
+
+		// push to active window, if it exists
+		if (this.current) {
+			this.current.casts.push(event)
+		}
+	}
+
+	private onApplyBuff(event: BuffEvent) {
+		// open a new window
+		this.startNewBuffWindow(event.timestamp)
+	}
+
+	private onRemoveBuff(event: BuffEvent) {
+		if (this.current) {
+			this.current.end = event.timestamp
+			this.windows.push(this.current)
+			this.current = undefined
+		}
+	}
+
+	// this is a helper because onCast might need to make a new window due to pre-pull buffs?
+	private startNewBuffWindow(startTime: number) {
+		this.current = new BuffWindowHistory(this.data, startTime, this.lastGcdIsBad)
+	}
+
+	protected onComplete() {
+		if (this.current) {
+			this.current.end = this.parser.fight.end_time
+			this.windows.push(this.current)
+		}
+
+		const badBuffs = this.windows.filter(window => window.isBad).length
+		this.suggestions.add(new TieredSuggestion({
+			icon: this.buffAction.icon,
+			content: <Trans id="drg.buffwindow.suggestions.bad-gcd">
+				Avoid using <ActionLink {...this.buffAction} /> immediately after <ActionLink {...ACTIONS.CHAOS_THRUST}/> or <ActionLink {...ACTIONS.FULL_THRUST}/> in order to get the most possible damage out of each window.
+			</Trans>,
+			tiers: {
+				1: SEVERITY.MINOR,
+				3: SEVERITY.MEDIUM,
+			},
+			value: badBuffs,
+			why: <Trans id="drg.buffwindow.suggestions.bad-gcd.why">
+				{badBuffs} of your <ActionLink {...this.buffAction} /> windows started right after a standard combo finisher.
+			</Trans>,
+		}))
+	}
+
+	renderTable() {
+		if (this.windows.length > 0) {
+			const rotationTargets = [{
+				header: <Trans id="drg.buffs.gcd-count">GCDs</Trans>,
+				accessor: 'gcds',
+			}]
+			const notesData = [{
+				header: <Trans id="drg.buffs.bad-start">Optimal Start</Trans>,
+				accessor: 'start',
+			}]
+
+			const rotationData = this.windows.map(window => {
+				const targets: RotationTableTargetData = {}
+				const notes: RotationTableNotesMap = {}
+				const windowStart = window.start - this.parser.eventTimeOffset
+				const windowEnd = (window.end != null ? window.end : window.start) - this.parser.eventTimeOffset
+
+				targets.gcds = {
+					actual: window.gcds,
+					expected: this.expectedGcds,
+				}
+
+				notes.start = window.isBad ? <Icon name="x" color="red" /> : <Icon name="check" color="green" />
+
+				// partial windows?
+				return {
+					start: windowStart,
+					end: windowEnd,
+					targetsData: targets,
+					rotation: window.casts,
+					notesMap: notes,
+				}
+			})
+
+			return <Fragment>
+				<RotationTable
+					targets={rotationTargets}
+					data={rotationData}
+					onGoto={this.timeline.show}
+					notes={notesData}
+				/>
+			</Fragment>
+		} else {
+			return <Fragment>
+				<Message warning>
+					<Trans id="drg.buffs.unused"><ActionLink {...this.buffAction} /> was not used during this fight.</Trans>
+				</Message>
+			</Fragment>
+		}
+	}
+
+	output() {
```

Comment:
Same concern as BattleLitany.tsx - is there a reason for having output() just be a wrapper on another function?